ntp-daemon/src/config/peer.rs

```diff
@@ -91,12 +106,36 @@ impl NormalizedAddress {
     pub(crate) fn new_unchecked(value: &str) -> Self {
         Self {
             address: value.to_string(),
+            hardcoded_dns_resolve: vec![],
         }
     }
 
+    #[cfg(test)]
+    pub(crate) fn with_hardcoded_dns(
+        address: &str,
+        hardcoded_dns_resolve: Vec<SocketAddr>,
+    ) -> Self {
+        Self {
+            address: address.to_string(),
+            hardcoded_dns_resolve,
+        }
+    }
+
+    #[cfg(not(test))]
     pub async fn lookup_host(&self) -> std::io::Result<impl Iterator<Item = SocketAddr> + '_> {
         tokio::net::lookup_host(&self.address).await
     }
+
+    #[cfg(test)]
+    pub async fn lookup_host(&self) -> std::io::Result<impl Iterator<Item = SocketAddr> + '_> {
+        let addresses = if !self.hardcoded_dns_resolve.is_empty() {
+            self.hardcoded_dns_resolve.to_vec()
+        } else {
+            tokio::net::lookup_host(&self.address).await?.collect()
+        };
+
+        Ok(addresses.into_iter())
+    }
```

Comment:
Although I see how this came about, I feel kind of uncomfortable that this leaves us with a function we could effectively never exercise in (non-integration) tests.